compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -397,22 +397,45 @@ impl<'hir> LoweringContext<'_, 'hir> {
         }
     }
 
-    // If `cond` kind is `let`, returns `let`. Otherwise, wraps and returns `cond`
-    // in a temporary block.
-    fn manage_let_cond(&mut self, cond: &'hir hir::Expr<'hir>) -> &'hir hir::Expr<'hir> {
+    // Wraps a condition (i.e. `cond` in `if cond` or `while cond`) in a terminating scope
+    // so that temporaries created in the condition don't live beyond it.
+    fn wrap_cond_in_drop_scope(&mut self, cond: &'hir hir::Expr<'hir>) -> &'hir hir::Expr<'hir> {
         fn has_let_expr<'hir>(expr: &'hir hir::Expr<'hir>) -> bool {
             match expr.kind {
                 hir::ExprKind::Binary(_, lhs, rhs) => has_let_expr(lhs) || has_let_expr(rhs),
                 hir::ExprKind::Let(..) => true,
                 _ => false,
             }
         }
-        if has_let_expr(cond) {
-            cond
-        } else {
-            let reason = DesugaringKind::CondTemporary;
-            let span_block = self.mark_span_with_reason(reason, cond.span, None);
-            self.expr_drop_temps(span_block, cond, AttrVec::new())
+
+        // We have to take special care for `let` exprs in the condition, e.g. in
+        // `if let pat = val` or `if foo && let pat = val`, as we _do_ want `val` to live beyond the
+        // condition in this case.
+        //
+        // In order to mantain the drop behavior for the non `let` parts of the condition,
+        // we still wrap them in terminating scopes, e.g. `if foo && let pat = val` essentially
+        // gets transformed into `if { let _t = foo; _t } && let pat = val`
+        match cond.kind {
+            hir::ExprKind::Binary(
+                op @ Spanned { node: hir::BinOpKind::And | hir::BinOpKind::Or, .. },
+                lhs,
+                rhs,
+            ) if has_let_expr(cond) => {
+                let lhs = self.wrap_cond_in_drop_scope(lhs);
+                let rhs = self.wrap_cond_in_drop_scope(rhs);
+
+                self.arena.alloc(self.expr(
```

Comment:
Ah yeah that makes sense, I pushed a commit to lower the condition directly from AST -> HIR instead of lowering to HIR and then transforming to the new HIR.